### PR TITLE
Pass in IFormFileCollection into minimal hosting delegate

### LINKED ISSUE #34303: Support binding to IFormFile in minimal routing APIs
### Is your feature request related to a problem? Please describe.
Today we don't have any native support for binding to IFormFiles as part of the incoming request. This is convenient for applications that want to do file uploads.

### Describe the solution you'd like

```C#
app.MapGet("/upload", async (IFormFile file) =>
{
    var uploads = Path.Combine(uploadsPath, file.FileName);
    using var fileStream = File.OpenWrite(uploads);
    using var uploadStream = file.OpenReadStream();
    await uploadStream.CopyToAsync(fileStream);
});
```

### Additional context
- Should we support IFormFileCollection?
- Would we do matching by name? (yes)

Without this, it will look like this:

```C#
app.MapGet("/upload", async (HttpRequest req) =>
{
    if (!req.HasFormContentType)
    {
        return Results.BadRequest();
    }

    var form = await req.ReadFormAsync();
    var file = form.Files["file"];

    if (file is null)
    {
        return Results.BadRequest();
    }

    var uploads = Path.Combine(uploadsPath, file.FileName);
    using var fileStream = File.OpenWrite(uploads);
    using var uploadStream = file.OpenReadStream();
    await uploadStream.CopyToAsync(fileStream);

    return Results.NoContent();
});
```

### PR DESCRIPTION
**Pull request by jchannon**
This adds IFormFileCollection IFormFileCollection into minimal hosting delegate.

Fixes #34303 

NOTE: Tests not added as cannot get any IDE to work so happy for someone to take this commit and carry on the work

### CODE REVIEW

**Comment by davidfowl** [comment]
This adds support for IFormFileCollection but what about IFormFile?

**Comment by jchannon** [comment]
Didn't quite know what the nameof would be for the first item in the
collection

File: src/Http/Http.Extensions/src/RequestDelegateFactory.cs [hunk 1]
```diff
@@ -55,6 +55,7 @@ public static partial class RequestDelegateFactory
         private static readonly MemberExpression HttpResponseExpr = Expression.Property(HttpContextExpr, nameof(HttpContext.Response));
         private static readonly MemberExpression RequestAbortedExpr = Expression.Property(HttpContextExpr, nameof(HttpContext.RequestAborted));
         private static readonly MemberExpression UserExpr = Expression.Property(HttpContextExpr, nameof(HttpContext.User));
+        private static readonly MemberExpression FormFileExpr = Expression.Property(HttpContextExpr, nameof(HttpContext.Request.Form.Files));
```
Comment (davidfowl) [inline src/Http/Http.Extensions/src/RequestDelegateFactory.cs:58]: This isn't how we would do this. Reading the form body should be done asynchronously. We'd probably store state on the FactoryContext that describes this is a form body that needs files. Take a look at [BindParameterFromBody](https://github.com/dotnet/aspnetcore/blob/34155282df177259d9b928080cbda9ca15a3afb3/src/Http/Http.Extensions/src/RequestDelegateFactory.cs#L715) and [HandleRequestBodyAndCompileRequestDelegate](https://github.com/dotnet/aspnetcore/blob/34155282df177259d9b928080cbda9ca15a3afb3/src/Http/Http.Extensions/src/RequestDelegateFactory.cs#L473)

**Comment by davidfowl** [comment]
This change is a bit more complex than what's implemented here:
- We need to pre-read the form in this case asynchronously if IFormFile or IFormFileCollection is used. This is alot simpler than what we do for JSON bodies as we don't need to pass any state after the read but we do need to pre-read the body in this case.
- We need to add support to the endpoints API explorer for detecting IFormFile and IFormFileCollection in the request body for Open API support.

**Pull request closed without merging**